### DISCUSSION TITLE
App Extensions: Less is more

### DISCUSSION THREAD

**Discussion by claytongulick** [directus/directus#14381] [Feature Requests]
I'm working with a lot of pretty complex extensions that are application specific, and having a lot of difficulty with the current process.

I know that the DX for extensions is under active development, but I wanted to prompt a conversation and provide some feedback for what I'd love to see for the extension DX.

I think there are two "philosophies" when it comes to extension development:

1 - Opinionated. Meaning, provide a full set of cli tools, rollup config, directory structure, scaffolding, Vue dependencies, etc...
 - This is great for new folks and for extensions that are simple components that are "vue friendly"

2 - Unopinionated / vanilla. At the end of the day, it's all just javascript. Vue is good at supporting non-build type components as well as vanilla web components. 
 - This is good for complex use cases where a different build system is used, or integration with complex external libs, direct DOM stuff, etc...

I think it'd be great to support both scenarios.

I think the case for 1 is already supported pretty well, in the /extensions/[extension type] folder, with the dynamic rollup and serving option.

I'm proposing a simple mechanism to also support 2: add a /extensions/app route.

On startup, the server looks in extensions/app folder to see if there's an index.js. If there is, it's served. If not, a no-op stub is served.

This index.js file would have a very simple interface, it would export an array of objects that match the current extension spec:

```
[
{
	id: 'some-extension',
	name: 'Some Extension',
	icon: 'box',
	description: 'Be Super Fancy and Use This!',
	component: SuperFancyComponent,
	options: null,
	types: ['json'],
},
{
...
}
]
```

Where SuperFancyComponent can be a vue component or vanilla Web Component. This would allow developers to use any sort of build tool, design system, tooling or libraries of their choice - as long as they compile down to Web Components, which most frameworks are able to do these days.

In order to make the use case of vanilla components possible, I also propose setting an additional property, by default, of all app extension components, in addition to primaryKey etc... the new property would be "context" and it would contain a reference to the api, stores and collections.

Last, I propose one more property "values" which would be set with the result of a watch(...) on the values injectable. This would allow reactivity from a vanilla component via a simple setter.

Thanks!!

**Comment by rijkvanzanten** [comment]
The main difficulty in opinionated vs unopinionated here is that even in the unopinionated version, we have to have _some_ opinions in what the output build looks like, as the app needs to be able to read it and execute it in a known way 🤔 That's also why the "opinionated" build tools started existing, as having people configure buildchains in that opinionated way of the app to interact with became quite a DX problem. But yeah, that does have the side effect that the tooling/flow becomes more and more opinionated 🤔 Tricky!

**Reply by claytongulick** [reply]
Looking through the code, it seems like that "contract" between the application host and the extension is already pretty well defined and pleasantly simple.

The json above is the info the application needs to know in order to properly hoist the components, and the props that are set are the inputs the components receive.

I'm working on an implementation of this idea today, because the separate build-per-extension isn't working for us, and is causing a lot of bloat from duplicate deps.

This, in turn, is causing a race condition: https://github.com/directus/directus/discussions/11254#discussioncomment-3123168

My initial attempt to address it was to add some plugins to the internal rollup call so that the build is centralized https://github.com/directus/directus/pull/14340#issuecomment-1179638315 but I think that this is the wrong approach. Playing "whackamole" by adding more and more rollups plugins to the internal extension manager like I did in that PR doesn't make sense.

I think a better approach is to support the "unopinionated" interface in addition to the existing method.

 I'll submit a PR as a proof of concept, hopefully today.

**Reply by claytongulick** [reply]
@rijkvanzanten what do you think of https://github.com/directus/directus/pull/14421 ?

It's massively improved our DX with extension development. Adds support for using vanilla Web Components as app extensions, as well as using external build tools and minification - solves the race condition we were running into.

**Reply by rijkvanzanten** [reply]
Tapping in @nickrum who is our resident extensions expert (and is the one who designed the current setup) 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
